### PR TITLE
3.13 compatibility changes (removed use of deprecated features)

### PR DESCRIPTION
**Pull request by dralley**

### CODE REVIEW

**Comment by dralley** [comment]
This will be broken until 3.12.1 releases with https://github.com/pulp/pulpcore/pull/1262

File: requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-pulpcore>=3.11
+pulpcore>=3.12.1
```
Comment (mdellweg) [inline requirements.txt:1]: :+1:

File: pulp_container/app/models.py [hunk 2]
```diff
@@ -355,6 +363,14 @@ def finalize_new_version(self, new_version):
...
         validate_repo_version(new_version)
 
+    def on_new_version(self, version):
+        """
+        Called when new repository versions are created.
+        Args:
+            version: The new repository version.
+        """
+        super().on_new_version(version)
+
```
Comment (mdellweg) [inline pulp_container/app/models.py:373]: I think, we don't need this.

File: pulp_container/app/models.py [hunk 3]
```diff
@@ -295,6 +295,14 @@ def finalize_new_version(self, new_version):
...
```
Comment (mdellweg) [inline pulp_container/app/models.py:305]: I think, we don't need this.

**Review by mdellweg — COMMENTED** [review]
Have you checked, whether permissions on the distribution will survive this migration?

**Comment by dralley** [comment]
@mdellweg addressed

File: pulp_container/app/migrations/0019_DATA_distribution_model_swap.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+# Generated by Django 2.2.19 on 2021-03-19 17:42
+
+from django.db import migrations, models, transaction
+import django.db.models.deletion
+
+
+def migrate_data_from_old_model_to_new_model_up(apps, schema_editor):
+    """ Move objects from ContainerDistribution to NewContainerDistribution."""
+    AnsibleDistribution = apps.get_model('container', 'ContainerDistribution')
```
Comment (ipanova) [inline pulp_container/app/migrations/0019_DATA_distribution_model_swap.py:9]: s/AnsibleDistribution/ContainerDistribution/

File: pulp_container/app/migrations/0019_DATA_distribution_model_swap.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            NewAnsibleDistribution(
+                pulp_id=container_distribution.pulp_id,
+                pulp_created=container_distribution.pulp_created,
+                pulp_last_updated=container_distribution.pulp_last_updated,
+                pulp_type=container_distribution.pulp_type,
+                name=container_distribution.name,
+                base_path=container_distribution.base_path,
+                content_guard=container_distribution.content_guard,
+                remote=container_distribution.remote,
+                repository_version=container_distribution.repository_version
```
Comment (ipanova) [inline pulp_container/app/migrations/0019_DATA_distribution_model_swap.py:22]: there are 3 additional fields on the Container distribution which also need data migration `description, namespace, private`

File: pulp_container/app/migrations/0019_DATA_distribution_model_swap.py (hunk 4 above)
Comment (dralley) [reply pulp_container/app/migrations/0019_DATA_distribution_model_swap.py]: How did I even miss that :/

**Comment by ipanova** [comment]
> Have you checked, whether permissions on the distribution will survive this migration?

yes
```
$ http  :24817/pulp/api/v3/groups/4/object_permissions/35/
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, DELETE, HEAD, OPTIONS
Connection: close
Content-Length: 215
Content-Type: application/json
Correlation-ID: d0e6f162563641f2b89e41c5d511c24b
Date: Wed, 21 Apr 2021 16:28:45 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "id": 35,
    "obj": "/pulp/api/v3/distributions/container/container/e402c32c-37fb-4182-871b-021ff9ff2543/",
    "permission": "container.delete_containerdistribution",
    "pulp_href": "/pulp/api/v3/groups/4/object_permissions/35/"
```

File: pulp_container/app/migrations/0019_DATA_distribution_model_swap.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            NewContainerDistribution(
+                pulp_id=container_distribution.pulp_id,
+                pulp_created=container_distribution.pulp_created,
+                pulp_last_updated=container_distribution.pulp_last_updated,
+                pulp_type=container_distribution.pulp_type,
+                name=container_distribution.name,
+                base_path=container_distribution.base_path,
+                content_guard=container_distribution.content_guard,
+                remote=container_distribution.remote,
+                repository_version=container_distribution.repository_version,
```
Comment (ipanova) [inline pulp_container/app/migrations/0019_DATA_distribution_model_swap.py:22]: we also need to migrate `repository`

**Comment by mdellweg** [comment]
> > Have you checked, whether permissions on the distribution will survive this migration?
> 
> yes

Hooray!!!

File: pulp_container/app/viewsets.py [hunk 7]
```diff
@@ -1299,19 +1313,21 @@ def destroy(self, request, pk, **kwargs):
...
+                    (
+                        str(distribution.repository.pk),
+                        "container",
+                        "ContainerPushRepositorySerializer",
+                    ),
                 )
 
         reservations.append(namespace)
         instance_ids.append(
             (namespace.pk, "container", "ContainerNamespaceSerializer"),
```
Comment (ipanova) [inline pulp_container/app/viewsets.py:1330]: str(namespace.pk)

**Review by ipanova — APPROVED** [review]
@dralley thank you for this PR. There is one last comment to be addressed https://github.com/pulp/pulp_container/pull/291#discussion_r617717753

**Comment by ipanova** [comment]
@mdellweg i could use another pair of eyes, if you don't find anything feel free to press merge button.

File: pulp_container/app/migrations/0019_DATA_distribution_model_swap.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+# Generated by Django 2.2.19 on 2021-03-19 17:42
```
Comment (mdellweg) [inline pulp_container/app/migrations/0019_DATA_distribution_model_swap.py:1]: The only thing i found was that this comment gives wrong credit. Should we remove that?

**Review by mdellweg — CHANGES_REQUESTED** [review]
All dispatch instances seem to have been addressed.
But i found leftover `enqueue_with_reservation` in the `registry_api.py`

**Review by mdellweg — APPROVED** [review]

**Comment by ipanova** [comment]
@dralley can you incorporate these changes so the tests pass? https://github.com/pulp/pulp_container/pull/299/commits/09084cabfabdf35e4611c3490ea58f36fca837b1

**Pull request closed without merging**